internal/api/core/route.go

```diff
@@ -0,0 +1,109 @@
+package core
+
+import (
+	"net/http"
+	"reflect"
+	"regexp"
+
+	"github.com/edulinq/autograder/internal/log"
+)
+
+type Route interface {
+	GetMethod() string
+	GetRegex() *regexp.Regexp
+	GetBasePath() string
+	Handle(response http.ResponseWriter, request *http.Request) error
+}
+
+// Inspired by https://benhoyt.com/writings/go-routing/
+type BaseRoute struct {
+	Method   string
+	BasePath string
+	Regex    *regexp.Regexp
+	Handler  RouteHandler
+}
+
+type APIRoute struct {
+	BaseRoute
+	RequestType  reflect.Type
+	ResponseType reflect.Type
+}
+
+func (this *BaseRoute) GetMethod() string {
+	return this.Method
+}
+
+func (this *BaseRoute) GetRegex() *regexp.Regexp {
+	return this.Regex
+}
+
+func (this *BaseRoute) GetBasePath() string {
+	return this.BasePath
+}
+
+func (this *BaseRoute) Handle(response http.ResponseWriter, request *http.Request) error {
+	return this.Handler(response, request)
+}
+
+func NewBaseRoute(method string, basePath string, handler RouteHandler) *BaseRoute {
+	return &BaseRoute{
+		Method:   method,
+		BasePath: basePath,
+		Regex:    regexp.MustCompile("^" + NewEndpoint(basePath) + "$"),
+		Handler:  handler,
+	}
+}
+
+func NewRedirect(method string, basePath string, target string) *BaseRoute {
+	redirectFunc := func(response http.ResponseWriter, request *http.Request) error {
+		return handleRedirect(target, response, request)
+	}
+
+	return &BaseRoute{
+		Method:   method,
+		BasePath: basePath,
+		Regex:    regexp.MustCompile("^" + NewEndpoint(basePath) + "$"),
+		Handler:  redirectFunc,
+	}
+}
+
+func NewAPIRoute(basePath string, apiHandler any) *APIRoute {
+	handler := func(response http.ResponseWriter, request *http.Request) (err error) {
+		// Recover from any panic.
+		defer func() {
+			value := recover()
+			if value == nil {
+				return
+			}
+
+			log.Error("Recovered from a panic when handling an API endpoint.",
+				log.NewAttr("value", value), log.NewAttr("endpoint", request.URL.Path))
+			apiErr := NewBareInternalError("-001", request.URL.Path, "Recovered from a panic when handling an API endpoint.").
+				Add("value", value)
+
+			err = sendAPIResponse(nil, response, nil, apiErr, false)
+		}()
+
+		err = handleAPIEndpoint(response, request, apiHandler)
+
+		return err
+	}
+
+	fullPath := NewEndpoint(basePath)
```

Comment:
1) I think we have an opportunity to make this function unexported
(which would be cool, since we generally don't want people mucking around with other paths).

2) We can also use this as a chance to change it's name to something better (maybe something like "makeFullAPIPath()"?).

3) This would also involve changing some test things (since the tests are using it).
I'm thinking that you could change SendTestAPIRequest*() to accept a basePath instead of the full path and then just build to full path in the test function.
(So we would pass something like `users/list` instead of `/api/v03/users/list`.)